File: Graph/python/islandCount.py
from .graph_utils import Grid, in_bounds, pos_key


def island_count(grid: Grid) -> int:
    visited: set[str] = set()
    count = 0

    for r in range(len(grid)):
        for c in range(len(grid[0])):
            if _explore(grid, r, c, visited):
                count += 1
    return count


def _explore(grid: Grid, r: int, c: int, visited: set[str]) -> bool:
    if not in_bounds(grid, r, c):
        return False
    if grid[r][c] == "W":
        return False

    key = pos_key(r, c)
    if key in visited:
        return False
    visited.add(key)

    _explore(grid, r - 1, c, visited)
    _explore(grid, r + 1, c, visited)
    _explore(grid, r, c - 1, visited)
    _explore(grid, r, c + 1, visited)
    return True

File: Graph/csharp/IslandCount.cs
using System.Collections.Generic;

namespace GraphSolutions;

public static class IslandCount
{
    public static int Solve(char[][] grid)
    {
        var visited = new HashSet<(int, int)>();
        var count = 0;

        for (var r = 0; r < grid.Length; r++)
        {
            for (var c = 0; c < grid[0].Length; c++)
            {
                if (Explore(grid, r, c, visited))
                {
                    count++;
                }
            }
        }

        return count;
    }

    private static bool Explore(char[][] grid, int row, int col, HashSet<(int, int)> visited)
    {
        if (!GraphUtils.InBounds(grid, row, col) || grid[row][col] == 'W')
        {
            return false;
        }

        var pos = (row, col);
        if (!visited.Add(pos))
        {
            return false;
        }

        foreach (var (dr, dc) in GraphUtils.Directions)
        {
            Explore(grid, row + dr, col + dc, visited);
        }

        return true;
    }
}
